test/contractBridge/whitelist.js

```diff
@@ -0,0 +1,172 @@
+        await TruffleAssert.reverts(BridgeInstance.disableWhitelist({from: genericAccount}),
+            "Sender doesn't have Whitelister or Admin role")
+    });
+
+    it('should allow admin to disable whitelist', async () => {
+        await BridgeInstance.enableWhitelist();
+        await BridgeInstance.grantRole(ADMIN_ROLE, genericAccount);
+        await TruffleAssert.passes(BridgeInstance.disableWhitelist({from: genericAccount}))
+    });
+
+    it('should allow whitelister to disable whitelist', async () => {
+        await BridgeInstance.enableWhitelist();
+        await BridgeInstance.grantRole(WHITELISTER_ROLE, genericAccount);
+        await TruffleAssert.passes(BridgeInstance.disableWhitelist({from: genericAccount}))
+    });
+
+    it('should require admin or whitelist role to add to whitelist', async () => {
+        await TruffleAssert.reverts(BridgeInstance.addToWhitelist(testAccount, {from: genericAccount}),
+            "Sender doesn't have Whitelister or Admin role")
+    });
+
+    it('should allow admin to add to whitelist', async () => {
+        await BridgeInstance.grantRole(ADMIN_ROLE, genericAccount);
+        await TruffleAssert.passes(BridgeInstance.addToWhitelist(testAccount, {from: genericAccount}))
+    });
+
+    it('should allow whitelister to add to whitelist', async () => {
+        await BridgeInstance.grantRole(WHITELISTER_ROLE, genericAccount);
+        await TruffleAssert.passes(BridgeInstance.addToWhitelist(testAccount, {from: genericAccount}))
+    });
+
+    it('should require admin or whitelist role to remove from whitelist', async () => {
+        await TruffleAssert.reverts(BridgeInstance.removeFromWhitelist(testAccount, {from: genericAccount}),
+            "Sender doesn't have Whitelister or Admin role")
+    });
+
+    it('should allow admin to remove from whitelist', async () => {
+        await BridgeInstance.addToWhitelist(testAccount);
+        await BridgeInstance.grantRole(ADMIN_ROLE, genericAccount);
+        await TruffleAssert.passes(BridgeInstance.removeFromWhitelist(testAccount, {from: genericAccount}))
+    });
+
+    it('should allow whitelister to remove from whitelist', async () => {
+        await BridgeInstance.addToWhitelist(testAccount);
+        await BridgeInstance.grantRole(WHITELISTER_ROLE, genericAccount);
+        await TruffleAssert.passes(BridgeInstance.removeFromWhitelist(testAccount, {from: genericAccount}))
+    });
+
+    it('should revert if whitelist is already enabled', async () => {
+        await BridgeInstance.enableWhitelist();
+        await TruffleAssert.reverts(BridgeInstance.enableWhitelist(),
+            "Whitelist is already enabled")
+    });
+
+    it('should revert if whitelist is already disabled', async () => {
+        await TruffleAssert.reverts(BridgeInstance.disableWhitelist(),
+            "Whitelist is already disabled")
+    });
+
+    it('should revert if address is already added to whitelist', async () => {
```

Comment:
This behavior is not really required. It’s easier to use an idempotent function because one can just call it to ensure the result even if it was there already.
That’s good too though, let’s keep it that way.

Reply:
`revert` helps to save gas:

![image](https://user-images.githubusercontent.com/29141708/117978445-90939c00-b33a-11eb-9b82-22ae0335a8fc.png)

Reply:
It also prevents duplicate events

Reply:
Good.